cpu/stm32_common/periph/rtc.c

```diff
@@ -322,4 +329,257 @@ void ISR_NAME(void)
+    RTC->CNTL = counter_val & 0x0000ffff;         // CNT[15:0]
+    _rtc_exit_config_mode();
+}
+
+// (UnixTime = 00:00:00 01.01.1970 = JD0 = 2440588)
+#define JULIAN_DATE_BASE    2440588
+
+// Julian day number calculation - https://en.wikipedia.org/wiki/Julian_day
+static void _rtc_get_time_from_count(const uint32_t rtc_counter, struct tm *time)
+{
+    unsigned long t;
+    unsigned long t1, a, b, c, d, e, m;
+    int year = 0;
+    int mon = 0;
+    int wday = 0;
+    int mday = 0;
+    int hour = 0;
+    int min = 0;
+    int sec = 0;
+    uint64_t jd = 0;;
+    uint64_t jdn = 0;
+
+    jd = ((rtc_counter + 43200) / (86400 >> 1)) + (2440587 <<1 ) + 1;
+    jdn = jd >> 1;
+
+    t = rtc_counter;
+    t1 = t / 60;
+    sec = t - t1 * 60;
+
+    t = t1;
+    t1 = t / 60;
+    min = t - t1 * 60;
+
+    t = t1;
+    t1 = t / 24;
+    hour = t - t1 * 24;
+
+    wday = jdn % 7;
+
+    a = jdn + 32044;
+    b = (4 * a + 3) / 146097;
+    c = a - (146097 * b) / 4;
+    d = (4 * c + 3) / 1461;
+    e = c - (1461 * d) / 4;
+    m = (5 * e + 2) / 153;
+    mday = e - (153 * m + 2) / 5 + 1;
+    mon = m + 3 - 12 * (m / 10);
+    year = 100 * b + d - 4800 + (m / 10);
+
+    time->tm_year = year - 1900;
+    time->tm_mon = mon;
+    time->tm_mday = mday;
+    time->tm_hour = hour;
+    time->tm_min = min;
+    time->tm_sec = sec;
+    time->tm_wday = wday;
+
+    DEBUG("[RTC time_from_count] c=%lu; %04i-%02i-%02i %02i:%02i:%02i\n", rtc_counter,
+        time->tm_year + 1900, time->tm_mon + 1, time->tm_mday,
+        time->tm_hour, time->tm_min, time->tm_sec);
+}
+
+// Julian day number calculation - https://en.wikipedia.org/wiki/Julian_day
+uint32_t _rtc_get_count_from_time(struct tm *time)
+{
+    uint8_t a;
+    uint16_t y;
+    uint8_t m;
+    uint32_t result_count;
+
+    a = (14 - time->tm_mon) / 12;
+    y = (time->tm_year + 1900) + 4800 - a;
+    m = time->tm_mon + (12 * a) - 3;
+
+    result_count = time->tm_mday;
+    result_count += (153 * m + 2) / 5;
+    result_count += 365 * y;
+    result_count += y / 4;
+    result_count += -y / 100;
+    result_count += y / 400;
+    result_count = result_count - 32045;
+    result_count = result_count - JULIAN_DATE_BASE;
+    result_count *= 86400;
+    result_count += (time->tm_hour * 3600);
+    result_count += (time->tm_min * 60);
+    result_count += (time->tm_sec);
+
+    DEBUG("[RTC count_from_time] c=%lu; %04i-%02i-%02i %02i:%02i:%02i\n", result_count,
+        time->tm_year + 1900, time->tm_mon + 1, time->tm_mday,
+        time->tm_hour, time->tm_min, time->tm_sec);
+
+    return result_count;
+}
+
+int rtc_set_time(struct tm *time)
+{
+    _rtc_set_counter_val(_rtc_get_count_from_time(time));
+    return 0;
+}
+
+int rtc_get_time(struct tm *time)
+{
+    uint32_t time_count_val = rtc_get_counter_val();
+    _rtc_get_time_from_count(time_count_val, time);
+    return 0;
+}
+
+int rtc_set_alarm(struct tm *time, rtc_alarm_cb_t cb, void *arg)
```

Comment:
This cannot be working, as both `cb` and `arg` are never used. The idea is that on the alarm `cb(arg)` is called. To do that, I would expect that both the function pointer and the argument are stored somewhere.

Have a lot at `struct isr_ctx;`, which is used by the other code and should be used here as well. Also the interrupt service routine `ISRNAME()` (yes, it is really named like that in the code) needs to be executed when the alarm is triggered. I didn't checked into the datasheet/documentation, so I have no idea how the isr needs to be registered.

Reply:
I use RTC only for wake up after turn off. Not use callback.
Now callback work fine.